// @pagarme/pagarme-nodejs-sdk/dist/esm/models/createDebitCardPaymentRequest.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { boolean, lazy, object, optional, string } from '../schema';
import { createCardPaymentContactlessRequestSchema, } from './createCardPaymentContactlessRequest';
import { createCardRequestSchema, } from './createCardRequest';
import { createPaymentAuthenticationRequestSchema, } from './createPaymentAuthenticationRequest';
import { createPaymentOriginRequestSchema, } from './createPaymentOriginRequest';
export const createDebitCardPaymentRequestSchema = object({
    statementDescriptor: ['statement_descriptor', optional(string())],
    card: ['card', optional(lazy(() => createCardRequestSchema))],
    cardId: ['card_id', optional(string())],
    cardToken: ['card_token', optional(string())],
    recurrence: ['recurrence', optional(boolean())],
    authentication: [
        'authentication',
        optional(lazy(() => createPaymentAuthenticationRequestSchema)),
    ],
    token: [
        'token',
        optional(lazy(() => createCardPaymentContactlessRequestSchema)),
    ],
    initiatedType: ['initiated_type', optional(string())],
    recurrenceModel: ['recurrence_model', optional(string())],
    paymentOrigin: [
        'payment_origin',
        optional(lazy(() => createPaymentOriginRequestSchema)),
    ],
});
