// @pagarme/pagarme-nodejs-sdk/dist/esm/models/createManagingPartnerRequest.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { array, bigint, boolean, lazy, nullable, object, optional, string, } from '../schema';
import { createRegisterInformationAddressRequestSchema, } from './createRegisterInformationAddressRequest';
import { createRegisterInformationPhoneRequestSchema, } from './createRegisterInformationPhoneRequest';
export const createManagingPartnerRequestSchema = object({
    name: ['name', string()],
    email: ['email', string()],
    document: ['document', string()],
    motherName: ['mother_name', optional(nullable(string()))],
    birthdate: ['birthdate', string()],
    monthlyIncome: ['monthly_income', bigint()],
    professionalOccupation: ['professional_occupation', string()],
    selfDeclaredLegalRepresentative: [
        'self_declared_legal_representative',
        boolean(),
    ],
    address: [
        'address',
        lazy(() => createRegisterInformationAddressRequestSchema),
    ],
    phoneNumbers: [
        'phone_numbers',
        array(lazy(() => createRegisterInformationPhoneRequestSchema)),
    ],
});
