// @pagarme/pagarme-nodejs-sdk/dist/esm/models/createCreditCardPaymentRequest.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { bigint, boolean, lazy, number, object, optional, string, } from '../schema';
import { createCardPayloadRequestSchema, } from './createCardPayloadRequest';
import { createCardPaymentContactlessRequestSchema, } from './createCardPaymentContactlessRequest';
import { createCardRequestSchema, } from './createCardRequest';
import { createPaymentAuthenticationRequestSchema, } from './createPaymentAuthenticationRequest';
import { createPaymentOriginRequestSchema, } from './createPaymentOriginRequest';
export const createCreditCardPaymentRequestSchema = object({
    installments: ['installments', optional(number())],
    statementDescriptor: ['statement_descriptor', optional(string())],
    card: ['card', optional(lazy(() => createCardRequestSchema))],
    cardId: ['card_id', optional(string())],
    cardToken: ['card_token', optional(string())],
    recurrence: ['recurrence', optional(boolean())],
    capture: ['capture', optional(boolean())],
    extendedLimitEnabled: ['extended_limit_enabled', optional(boolean())],
    extendedLimitCode: ['extended_limit_code', optional(string())],
    merchantCategoryCode: ['merchant_category_code', optional(bigint())],
    authentication: [
        'authentication',
        optional(lazy(() => createPaymentAuthenticationRequestSchema)),
    ],
    contactless: [
        'contactless',
        optional(lazy(() => createCardPaymentContactlessRequestSchema)),
    ],
    autoRecovery: ['auto_recovery', optional(boolean())],
    operationType: ['operation_type', optional(string())],
    recurrencyCycle: ['recurrency_cycle', optional(string())],
    payload: ['payload', optional(lazy(() => createCardPayloadRequestSchema))],
    initiatedType: ['initiated_type', optional(string())],
    recurrenceModel: ['recurrence_model', optional(string())],
    paymentOrigin: [
        'payment_origin',
        optional(lazy(() => createPaymentOriginRequestSchema)),
    ],
});
