// @pagarme/pagarme-nodejs-sdk/dist/esm/models/createBoletoPaymentRequest.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { lazy, nullable, number, object, optional, string, } from '../schema';
import { createAddressRequestSchema, } from './createAddressRequest';
import { createFineRequestSchema, } from './createFineRequest';
import { createInterestRequestSchema, } from './createInterestRequest';
export const createBoletoPaymentRequestSchema = object({
    retries: ['retries', number()],
    bank: ['bank', optional(nullable(string()))],
    instructions: ['instructions', string()],
    dueAt: ['due_at', optional(nullable(string()))],
    billingAddress: ['billing_address', lazy(() => createAddressRequestSchema)],
    billingAddressId: ['billing_address_id', optional(nullable(string()))],
    nossoNumero: ['nosso_numero', optional(nullable(string()))],
    documentNumber: ['document_number', string()],
    statementDescriptor: ['statement_descriptor', string()],
    interest: [
        'interest',
        optional(nullable(lazy(() => createInterestRequestSchema))),
    ],
    fine: ['fine', optional(nullable(lazy(() => createFineRequestSchema)))],
    maxDaysToPayPastDue: [
        'max_days_to_pay_past_due',
        optional(nullable(number())),
    ],
});
