// @pagarme/pagarme-nodejs-sdk/dist/esm/models/createChargeRequest.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { dict, lazy, nullable, number, object, optional, string, } from '../schema';
import { createAntifraudRequestSchema, } from './createAntifraudRequest';
import { createCustomerRequestSchema, } from './createCustomerRequest';
import { createPaymentRequestSchema, } from './createPaymentRequest';
export const createChargeRequestSchema = object({
    code: ['code', optional(nullable(string()))],
    amount: ['amount', number()],
    customerId: ['customer_id', optional(nullable(string()))],
    customer: [
        'customer',
        optional(nullable(lazy(() => createCustomerRequestSchema))),
    ],
    payment: ['payment', lazy(() => createPaymentRequestSchema)],
    metadata: ['metadata', optional(nullable(dict(string())))],
    dueAt: ['due_at', optional(nullable(string()))],
    antifraud: [
        'antifraud',
        optional(nullable(lazy(() => createAntifraudRequestSchema))),
    ],
    orderId: ['order_id', string()],
});
