// @pagarme/pagarme-nodejs-sdk/dist/esm/models/createRegisterInformationIndividualRequest.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { array, bigint, lazy, nullable, object, optional, string, } from '../schema';
import { createRegisterInformationAddressRequestSchema, } from './createRegisterInformationAddressRequest';
import { createRegisterInformationPhoneRequestSchema } from './createRegisterInformationPhoneRequest';
export const createRegisterInformationIndividualRequestSchema = object({
    name: ['name', string()],
    motherName: ['mother_name', optional(nullable(string()))],
    birthdate: ['birthdate', string()],
    monthlyIncome: ['monthly_income', bigint()],
    professionalOccupation: ['professional_occupation', string()],
    address: [
        'address',
        lazy(() => createRegisterInformationAddressRequestSchema),
    ],
    email: ['email', string()],
    document: ['document', string()],
    type: ['type', string()],
    siteUrl: ['site_url', optional(nullable(string()))],
    phoneNumbers: [
        'phone_numbers',
        array(lazy(() => createRegisterInformationPhoneRequestSchema)),
    ],
});
