// @pagarme/pagarme-nodejs-sdk/dist/esm/models/createRecipientRequest.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { dict, lazy, nullable, object, optional, string, } from '../schema';
import { createBankAccountRequestSchema, } from './createBankAccountRequest';
import { createRegisterInformationBaseRequestSchema, } from './createRegisterInformationBaseRequest';
import { createTransferSettingsRequestSchema, } from './createTransferSettingsRequest';
export const createRecipientRequestSchema = object({
    name: ['name', optional(nullable(string()))],
    email: ['email', optional(nullable(string()))],
    description: ['description', optional(nullable(string()))],
    document: ['document', optional(nullable(string()))],
    type: ['type', optional(nullable(string()))],
    defaultBankAccount: [
        'default_bank_account',
        lazy(() => createBankAccountRequestSchema),
    ],
    metadata: ['metadata', dict(string())],
    transferSettings: [
        'transfer_settings',
        optional(lazy(() => createTransferSettingsRequestSchema)),
    ],
    code: ['code', string()],
    paymentMode: ['payment_mode', string()],
    registerInformation: [
        'register_information',
        optional(nullable(lazy(() => createRegisterInformationBaseRequestSchema))),
    ],
});
