// @pagarme/pagarme-nodejs-sdk/dist/esm/models/createCardRequest.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { boolean, dict, lazy, number, object, optional, string, } from '../schema';
import { createAddressRequestSchema, } from './createAddressRequest';
import { createCardOptionsRequestSchema, } from './createCardOptionsRequest';
export const createCardRequestSchema = object({
    number: ['number', optional(string())],
    holderName: ['holder_name', optional(string())],
    expMonth: ['exp_month', optional(number())],
    expYear: ['exp_year', optional(number())],
    cvv: ['cvv', optional(string())],
    billingAddress: [
        'billing_address',
        optional(lazy(() => createAddressRequestSchema)),
    ],
    brand: ['brand', optional(string())],
    billingAddressId: ['billing_address_id', optional(string())],
    metadata: ['metadata', optional(dict(string()))],
    type: ['type', optional(string())],
    options: ['options', optional(lazy(() => createCardOptionsRequestSchema))],
    holderDocument: ['holder_document', optional(string())],
    privateLabel: ['private_label', optional(boolean())],
    label: ['label', optional(string())],
    id: ['id', optional(string())],
    token: ['token', optional(string())],
});
