// @pagarme/pagarme-nodejs-sdk/dist/esm/models/createPlanRequest.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { array, boolean, dict, lazy, number, object, optional, string, } from '../schema';
import { createPlanItemRequestSchema, } from './createPlanItemRequest';
import { createPricingSchemeRequestSchema, } from './createPricingSchemeRequest';
export const createPlanRequestSchema = object({
    name: ['name', string()],
    description: ['description', string()],
    statementDescriptor: ['statement_descriptor', string()],
    items: ['items', array(lazy(() => createPlanItemRequestSchema))],
    shippable: ['shippable', boolean()],
    paymentMethods: ['payment_methods', array(string())],
    installments: ['installments', array(number())],
    currency: ['currency', string()],
    interval: ['interval', string()],
    intervalCount: ['interval_count', number()],
    billingDays: ['billing_days', array(number())],
    billingType: ['billing_type', string()],
    pricingScheme: [
        'pricing_scheme',
        lazy(() => createPricingSchemeRequestSchema),
    ],
    metadata: ['metadata', dict(string())],
    minimumPrice: ['minimum_price', optional(number())],
    cycles: ['cycles', optional(number())],
    quantity: ['quantity', optional(number())],
    trialPeriodDays: ['trial_period_days', optional(number())],
});
