// @pagarme/pagarme-nodejs-sdk/dist/esm/models/createRegisterInformationCorporationRequest.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { array, bigint, lazy, nullable, object, optional, string, } from '../schema';
import { createManagingPartnerRequestSchema, } from './createManagingPartnerRequest';
import { createRegisterInformationAddressRequestSchema, } from './createRegisterInformationAddressRequest';
import { createRegisterInformationPhoneRequestSchema } from './createRegisterInformationPhoneRequest';
export const createRegisterInformationCorporationRequestSchema = object({
    companyName: ['company_name', string()],
    tradingName: ['trading_name', string()],
    annualRevenue: ['annual_revenue', bigint()],
    corporationType: ['corporation_type', optional(nullable(string()))],
    foundingDate: ['founding_date', optional(nullable(string()))],
    cnae: ['cnae', optional(nullable(string()))],
    managingPartners: [
        'managing_partners',
        array(lazy(() => createManagingPartnerRequestSchema)),
    ],
    mainAddress: [
        'main_address',
        lazy(() => createRegisterInformationAddressRequestSchema),
    ],
    email: ['email', string()],
    document: ['document', string()],
    type: ['type', string()],
    siteUrl: ['site_url', optional(nullable(string()))],
    phoneNumbers: [
        'phone_numbers',
        array(lazy(() => createRegisterInformationPhoneRequestSchema)),
    ],
});
