// @pagarme/pagarme-nodejs-sdk/dist/esm/models/getAddressResponse.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { dict, lazy, nullable, object, optional, string, } from '../schema';
import { getCustomerResponseSchema, } from './getCustomerResponse';
export const getAddressResponseSchema = object({
    id: ['id', optional(nullable(string()))],
    street: ['street', optional(nullable(string()))],
    number: ['number', optional(nullable(string()))],
    complement: ['complement', optional(nullable(string()))],
    zipCode: ['zip_code', optional(nullable(string()))],
    neighborhood: ['neighborhood', optional(nullable(string()))],
    city: ['city', optional(nullable(string()))],
    state: ['state', optional(nullable(string()))],
    country: ['country', optional(nullable(string()))],
    status: ['status', optional(nullable(string()))],
    createdAt: ['created_at', optional(nullable(string()))],
    updatedAt: ['updated_at', optional(nullable(string()))],
    customer: [
        'customer',
        optional(nullable(lazy(() => getCustomerResponseSchema))),
    ],
    metadata: ['metadata', optional(nullable(dict(string())))],
    line1: ['line_1', optional(nullable(string()))],
    line2: ['line_2', optional(nullable(string()))],
    deletedAt: ['deleted_at', optional(nullable(string()))],
});
