// @pagarme/pagarme-nodejs-sdk/dist/esm/models/getBankAccountResponse.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { dict, lazy, nullable, object, optional, string, } from '../schema';
import { getRecipientResponseSchema, } from './getRecipientResponse';
export const getBankAccountResponseSchema = object({
    id: ['id', optional(nullable(string()))],
    holderName: ['holder_name', optional(nullable(string()))],
    holderType: ['holder_type', optional(nullable(string()))],
    bank: ['bank', optional(nullable(string()))],
    branchNumber: ['branch_number', optional(nullable(string()))],
    branchCheckDigit: ['branch_check_digit', optional(nullable(string()))],
    accountNumber: ['account_number', optional(nullable(string()))],
    accountCheckDigit: ['account_check_digit', optional(nullable(string()))],
    type: ['type', optional(nullable(string()))],
    status: ['status', optional(nullable(string()))],
    createdAt: ['created_at', optional(nullable(string()))],
    updatedAt: ['updated_at', optional(nullable(string()))],
    deletedAt: ['deleted_at', optional(nullable(string()))],
    recipient: [
        'recipient',
        optional(nullable(lazy(() => getRecipientResponseSchema))),
    ],
    metadata: ['metadata', optional(nullable(dict(string())))],
    pixKey: ['pix_key', optional(nullable(string()))],
});
