// @pagarme/pagarme-nodejs-sdk/dist/esm/models/getCheckoutPaymentSettingsResponse.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { array, lazy, nullable, number, object, optional, string, } from '../schema';
import { getCustomerResponseSchema, } from './getCustomerResponse';
export const getCheckoutPaymentSettingsResponseSchema = object({
    successUrl: ['success_url', optional(nullable(string()))],
    paymentUrl: ['payment_url', optional(nullable(string()))],
    acceptedPaymentMethods: [
        'accepted_payment_methods',
        optional(nullable(array(string()))),
    ],
    status: ['status', optional(nullable(string()))],
    customer: [
        'customer',
        optional(nullable(lazy(() => getCustomerResponseSchema))),
    ],
    amount: ['amount', optional(nullable(number()))],
    defaultPaymentMethod: [
        'default_payment_method',
        optional(nullable(string())),
    ],
    gatewayAffiliationId: [
        'gateway_affiliation_id',
        optional(nullable(string())),
    ],
});
