// @pagarme/pagarme-nodejs-sdk/dist/esm/models/getCustomerResponse.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { bigint, boolean, dict, lazy, nullable, object, optional, string, } from '../schema';
import { getAddressResponseSchema, } from './getAddressResponse';
import { getPhonesResponseSchema, } from './getPhonesResponse';
export const getCustomerResponseSchema = object({
    id: ['id', optional(nullable(string()))],
    name: ['name', optional(nullable(string()))],
    email: ['email', optional(nullable(string()))],
    delinquent: ['delinquent', optional(nullable(boolean()))],
    createdAt: ['created_at', optional(nullable(string()))],
    updatedAt: ['updated_at', optional(nullable(string()))],
    document: ['document', optional(nullable(string()))],
    type: ['type', optional(nullable(string()))],
    fbAccessToken: ['fb_access_token', optional(nullable(string()))],
    address: [
        'address',
        optional(nullable(lazy(() => getAddressResponseSchema))),
    ],
    metadata: ['metadata', optional(nullable(dict(string())))],
    phones: ['phones', optional(nullable(lazy(() => getPhonesResponseSchema)))],
    fbId: ['fb_id', optional(nullable(bigint()))],
    code: ['code', optional(nullable(string()))],
    documentType: ['document_type', optional(nullable(string()))],
});
