// @pagarme/pagarme-nodejs-sdk/dist/esm/models/getCardResponse.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { dict, lazy, nullable, number, object, optional, string, } from '../schema';
import { getBillingAddressResponseSchema, } from './getBillingAddressResponse';
import { getCustomerResponseSchema, } from './getCustomerResponse';
export const getCardResponseSchema = object({
    id: ['id', optional(nullable(string()))],
    lastFourDigits: ['last_four_digits', optional(nullable(string()))],
    brand: ['brand', optional(nullable(string()))],
    holderName: ['holder_name', optional(nullable(string()))],
    expMonth: ['exp_month', optional(nullable(number()))],
    expYear: ['exp_year', optional(nullable(number()))],
    status: ['status', optional(nullable(string()))],
    createdAt: ['created_at', optional(nullable(string()))],
    updatedAt: ['updated_at', optional(nullable(string()))],
    billingAddress: [
        'billing_address',
        optional(nullable(lazy(() => getBillingAddressResponseSchema))),
    ],
    customer: [
        'customer',
        optional(nullable(lazy(() => getCustomerResponseSchema))),
    ],
    metadata: ['metadata', optional(nullable(dict(string())))],
    type: ['type', optional(nullable(string()))],
    holderDocument: ['holder_document', optional(nullable(string()))],
    deletedAt: ['deleted_at', optional(nullable(string()))],
    firstSixDigits: ['first_six_digits', optional(nullable(string()))],
    label: ['label', optional(nullable(string()))],
});
