// @pagarme/pagarme-nodejs-sdk/dist/esm/models/getDiscountResponse.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { lazy, nullable, number, object, optional, string, } from '../schema';
import { getSubscriptionItemResponseSchema, } from './getSubscriptionItemResponse';
import { getSubscriptionResponseSchema, } from './getSubscriptionResponse';
export const getDiscountResponseSchema = object({
    id: ['id', optional(nullable(string()))],
    value: ['value', optional(nullable(number()))],
    discountType: ['discount_type', optional(nullable(string()))],
    status: ['status', optional(nullable(string()))],
    createdAt: ['created_at', optional(nullable(string()))],
    cycles: ['cycles', optional(nullable(number()))],
    deletedAt: ['deleted_at', optional(nullable(string()))],
    description: ['description', optional(nullable(string()))],
    subscription: [
        'subscription',
        optional(nullable(lazy(() => getSubscriptionResponseSchema))),
    ],
    subscriptionItem: [
        'subscription_item',
        optional(nullable(lazy(() => getSubscriptionItemResponseSchema))),
    ],
});
