// @pagarme/pagarme-nodejs-sdk/dist/esm/models/getInvoiceResponse.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { array, dict, lazy, nullable, number, object, optional, string, } from '../schema';
import { getBillingAddressResponseSchema, } from './getBillingAddressResponse';
import { getChargeResponseSchema, } from './getChargeResponse';
import { getCustomerResponseSchema, } from './getCustomerResponse';
import { getInvoiceItemResponseSchema, } from './getInvoiceItemResponse';
import { getPeriodResponseSchema, } from './getPeriodResponse';
import { getShippingResponseSchema, } from './getShippingResponse';
import { getSubscriptionResponseSchema, } from './getSubscriptionResponse';
export const getInvoiceResponseSchema = object({
    id: ['id', optional(nullable(string()))],
    code: ['code', optional(nullable(string()))],
    url: ['url', optional(nullable(string()))],
    amount: ['amount', optional(nullable(number()))],
    status: ['status', optional(nullable(string()))],
    paymentMethod: ['payment_method', optional(nullable(string()))],
    createdAt: ['created_at', optional(nullable(string()))],
    items: [
        'items',
        optional(nullable(array(lazy(() => getInvoiceItemResponseSchema)))),
    ],
    customer: [
        'customer',
        optional(nullable(lazy(() => getCustomerResponseSchema))),
    ],
    charge: ['charge', optional(nullable(lazy(() => getChargeResponseSchema)))],
    installments: ['installments', optional(nullable(number()))],
    billingAddress: [
        'billing_address',
        optional(nullable(lazy(() => getBillingAddressResponseSchema))),
    ],
    subscription: [
        'subscription',
        optional(nullable(lazy(() => getSubscriptionResponseSchema))),
    ],
    cycle: ['cycle', optional(nullable(lazy(() => getPeriodResponseSchema)))],
    shipping: [
        'shipping',
        optional(nullable(lazy(() => getShippingResponseSchema))),
    ],
    metadata: ['metadata', optional(nullable(dict(string())))],
    dueAt: ['due_at', optional(nullable(string()))],
    canceledAt: ['canceled_at', optional(nullable(string()))],
    billingAt: ['billing_at', optional(nullable(string()))],
    seenAt: ['seen_at', optional(nullable(string()))],
    totalDiscount: ['total_discount', optional(nullable(number()))],
    totalIncrement: ['total_increment', optional(nullable(number()))],
    subscriptionId: ['subscription_id', optional(nullable(string()))],
});
