// @pagarme/pagarme-nodejs-sdk/dist/esm/models/getMovementObjectRefundResponse.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { nullable, object, optional, string } from '../schema';
export const getMovementObjectRefundResponseSchema = object({
    fraudCoverageFee: ['fraud_coverage_fee', optional(nullable(string()))],
    chargeFeeRecipientId: [
        'charge_fee_recipient_id',
        optional(nullable(string())),
    ],
    bankAccountId: ['bank_account_id', optional(nullable(string()))],
    localTransactionId: ['local_transaction_id', optional(nullable(string()))],
    updatedAt: ['updated_at', optional(nullable(string()))],
    id: ['id', optional(nullable(string()))],
    status: ['status', optional(nullable(string()))],
    amount: ['amount', optional(nullable(string()))],
    createdAt: ['created_at', optional(nullable(string()))],
    type: ['type', optional(nullable(string()))],
    chargeId: ['charge_id', optional(nullable(string()))],
    gatewayId: ['gateway_id', optional(nullable(string()))],
    object: ['object', optional(string())],
});
