// @pagarme/pagarme-nodejs-sdk/dist/esm/models/getMovementObjectSettlementResponse.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { nullable, object, optional, string } from '../schema';
export const getMovementObjectSettlementResponseSchema = object({
    product: ['product', optional(nullable(string()))],
    brand: ['brand', optional(nullable(string()))],
    paymentDate: ['payment_date', optional(nullable(string()))],
    recipientId: ['recipient_id', optional(nullable(string()))],
    documentType: ['document_type', optional(nullable(string()))],
    document: ['document', optional(nullable(string()))],
    contractObligationId: [
        'contract_obligation_id',
        optional(nullable(string())),
    ],
    liquidationArrangementId: [
        'liquidation_arrangement_id',
        optional(nullable(string())),
    ],
    externalEnginePaymentId: [
        'external_engine_payment_id',
        optional(nullable(string())),
    ],
    id: ['id', optional(nullable(string()))],
    status: ['status', optional(nullable(string()))],
    amount: ['amount', optional(nullable(string()))],
    createdAt: ['created_at', optional(nullable(string()))],
    type: ['type', optional(nullable(string()))],
    chargeId: ['charge_id', optional(nullable(string()))],
    gatewayId: ['gateway_id', optional(nullable(string()))],
    object: ['object', optional(string())],
});
