// @pagarme/pagarme-nodejs-sdk/dist/esm/models/getManagingPartnerResponse.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { array, boolean, lazy, nullable, object, optional, string, } from '../schema';
import { getPhoneNumberResponseSchema, } from './getPhoneNumberResponse';
import { getRegisterInformationAddressResponseSchema, } from './getRegisterInformationAddressResponse';
export const getManagingPartnerResponseSchema = object({
    name: ['name', optional(nullable(string()))],
    email: ['email', optional(nullable(string()))],
    document: ['document', optional(nullable(string()))],
    type: ['type', optional(nullable(string()))],
    motherName: ['mother_name', optional(nullable(string()))],
    birthdate: ['birthdate', optional(nullable(string()))],
    monthlyIncome: ['monthly_income', optional(nullable(string()))],
    professionalOccupation: [
        'professional_occupation',
        optional(nullable(string())),
    ],
    selfDeclaredRepresentative: [
        'self_declared_representative',
        optional(boolean()),
    ],
    address: [
        'address',
        optional(nullable(lazy(() => getRegisterInformationAddressResponseSchema))),
    ],
    phoneNumbers: [
        'phone_numbers',
        optional(nullable(array(lazy(() => getPhoneNumberResponseSchema)))),
    ],
});
