// @pagarme/pagarme-nodejs-sdk/dist/esm/models/getPayableResponse.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { bigint, nullable, number, object, optional, string, } from '../schema';
export const getPayableResponseSchema = object({
    id: ['id', optional(nullable(bigint()))],
    status: ['status', optional(nullable(string()))],
    amount: ['amount', optional(nullable(number()))],
    fee: ['fee', optional(nullable(number()))],
    anticipationFee: ['anticipation_fee', optional(nullable(number()))],
    fraudCoverageFee: ['fraud_coverage_fee', optional(nullable(number()))],
    installment: ['installment', optional(nullable(number()))],
    gatewayId: ['gateway_id', optional(nullable(bigint()))],
    chargeId: ['charge_id', optional(nullable(string()))],
    splitId: ['split_id', optional(nullable(string()))],
    bulkAnticipationId: ['bulk_anticipation_id', optional(nullable(string()))],
    anticipationId: ['anticipation_id', optional(nullable(string()))],
    recipientId: ['recipient_id', optional(nullable(string()))],
    originatorModel: ['originator_model', optional(nullable(string()))],
    originatorModelId: ['originator_model_id', optional(nullable(string()))],
    paymentDate: ['payment_date', optional(nullable(string()))],
    originalPaymentDate: ['original_payment_date', optional(nullable(string()))],
    type: ['type', optional(nullable(string()))],
    paymentMethod: ['payment_method', optional(nullable(string()))],
    accrualAt: ['accrual_at', optional(nullable(string()))],
    createdAt: ['created_at', optional(nullable(string()))],
    liquidationArrangementId: [
        'liquidation_arrangement_id',
        optional(nullable(string())),
    ],
});
