// @pagarme/pagarme-nodejs-sdk/dist/esm/models/getPlanResponse.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { array, boolean, dict, lazy, nullable, number, object, optional, string, } from '../schema';
import { getPlanItemResponseSchema, } from './getPlanItemResponse';
export const getPlanResponseSchema = object({
    id: ['id', optional(nullable(string()))],
    name: ['name', optional(nullable(string()))],
    description: ['description', optional(nullable(string()))],
    url: ['url', optional(nullable(string()))],
    statementDescriptor: ['statement_descriptor', optional(nullable(string()))],
    interval: ['interval', optional(nullable(string()))],
    intervalCount: ['interval_count', optional(nullable(number()))],
    billingType: ['billing_type', optional(nullable(string()))],
    paymentMethods: ['payment_methods', optional(nullable(array(string())))],
    installments: ['installments', optional(nullable(array(number())))],
    status: ['status', optional(nullable(string()))],
    currency: ['currency', optional(nullable(string()))],
    createdAt: ['created_at', optional(nullable(string()))],
    updatedAt: ['updated_at', optional(nullable(string()))],
    items: [
        'items',
        optional(nullable(array(lazy(() => getPlanItemResponseSchema)))),
    ],
    billingDays: ['billing_days', optional(nullable(array(number())))],
    shippable: ['shippable', optional(nullable(boolean()))],
    metadata: ['metadata', optional(nullable(dict(string())))],
    trialPeriodDays: ['trial_period_days', optional(nullable(number()))],
    minimumPrice: ['minimum_price', optional(nullable(number()))],
    deletedAt: ['deleted_at', optional(nullable(string()))],
});
