// @pagarme/pagarme-nodejs-sdk/dist/esm/models/getPlanItemResponse.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { lazy, nullable, number, object, optional, string, } from '../schema';
import { getPlanResponseSchema } from './getPlanResponse';
import { getPricingSchemeResponseSchema, } from './getPricingSchemeResponse';
export const getPlanItemResponseSchema = object({
    id: ['id', optional(nullable(string()))],
    name: ['name', optional(nullable(string()))],
    status: ['status', optional(nullable(string()))],
    createdAt: ['created_at', optional(nullable(string()))],
    updatedAt: ['updated_at', optional(nullable(string()))],
    pricingScheme: [
        'pricing_scheme',
        optional(nullable(lazy(() => getPricingSchemeResponseSchema))),
    ],
    description: ['description', optional(nullable(string()))],
    plan: ['plan', optional(nullable(lazy(() => getPlanResponseSchema)))],
    quantity: ['quantity', optional(nullable(number()))],
    cycles: ['cycles', optional(nullable(number()))],
    deletedAt: ['deleted_at', optional(nullable(string()))],
});
