// @pagarme/pagarme-nodejs-sdk/dist/esm/models/getRecipientResponse.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { array, dict, lazy, nullable, object, optional, string, } from '../schema';
import { getAutomaticAnticipationResponseSchema, } from './getAutomaticAnticipationResponse';
import { getBankAccountResponseSchema, } from './getBankAccountResponse';
import { getGatewayRecipientResponseSchema, } from './getGatewayRecipientResponse';
import { getRegisterInformationResponseSchema, } from './getRegisterInformationResponse';
import { getTransferSettingsResponseSchema, } from './getTransferSettingsResponse';
export const getRecipientResponseSchema = object({
    id: ['id', optional(nullable(string()))],
    name: ['name', optional(nullable(string()))],
    email: ['email', optional(nullable(string()))],
    document: ['document', optional(nullable(string()))],
    description: ['description', optional(nullable(string()))],
    type: ['type', optional(nullable(string()))],
    status: ['status', optional(nullable(string()))],
    createdAt: ['created_at', optional(nullable(string()))],
    updatedAt: ['updated_at', optional(nullable(string()))],
    deletedAt: ['deleted_at', optional(nullable(string()))],
    defaultBankAccount: [
        'default_bank_account',
        optional(nullable(lazy(() => getBankAccountResponseSchema))),
    ],
    gatewayRecipients: [
        'gateway_recipients',
        optional(nullable(array(lazy(() => getGatewayRecipientResponseSchema)))),
    ],
    metadata: ['metadata', optional(nullable(dict(string())))],
    automaticAnticipationSettings: [
        'automatic_anticipation_settings',
        optional(nullable(lazy(() => getAutomaticAnticipationResponseSchema))),
    ],
    transferSettings: [
        'transfer_settings',
        optional(nullable(lazy(() => getTransferSettingsResponseSchema))),
    ],
    code: ['code', optional(nullable(string()))],
    paymentMode: ['payment_mode', optional(nullable(string()))],
    registerInformation: [
        'register_information',
        optional(nullable(lazy(() => getRegisterInformationResponseSchema))),
    ],
});
