// @pagarme/pagarme-nodejs-sdk/dist/esm/models/getRegisterInformationResponse.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { array, lazy, nullable, object, optional, string, } from '../schema';
import { getManagingPartnerResponseSchema, } from './getManagingPartnerResponse';
import { getPhoneNumberResponseSchema, } from './getPhoneNumberResponse';
import { getRegisterInformationAddressResponseSchema, } from './getRegisterInformationAddressResponse';
export const getRegisterInformationResponseSchema = object({
    email: ['email', optional(nullable(string()))],
    document: ['document', optional(nullable(string()))],
    type: ['type', optional(nullable(string()))],
    siteUrl: ['site_url', optional(nullable(string()))],
    phoneNumbers: [
        'phone_numbers',
        optional(nullable(array(lazy(() => getPhoneNumberResponseSchema)))),
    ],
    name: ['name', optional(nullable(string()))],
    motherName: ['mother_name', optional(nullable(string()))],
    birthdate: ['birthdate', optional(nullable(string()))],
    monthlyIncome: ['monthly_income', optional(nullable(string()))],
    professionalOccupation: [
        'professional_occupation',
        optional(nullable(string())),
    ],
    address: [
        'address',
        optional(nullable(lazy(() => getRegisterInformationAddressResponseSchema))),
    ],
    companyName: ['company_name', optional(nullable(string()))],
    tradingName: ['trading_name', optional(nullable(string()))],
    annualRevenue: ['annual_revenue', optional(nullable(string()))],
    corporationType: ['corporation_type', optional(nullable(string()))],
    foundingDate: ['founding_date', optional(nullable(string()))],
    cnae: ['cnae', optional(nullable(string()))],
    mainAddress: [
        'main_address',
        optional(nullable(lazy(() => getRegisterInformationAddressResponseSchema))),
    ],
    managingPartners: [
        'managing_partners',
        optional(nullable(array(lazy(() => getManagingPartnerResponseSchema)))),
    ],
});
