// @pagarme/pagarme-nodejs-sdk/dist/esm/models/getSubscriptionItemResponse.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { array, lazy, nullable, number, object, optional, string, } from '../schema';
import { getDiscountResponseSchema, } from './getDiscountResponse';
import { getIncrementResponseSchema, } from './getIncrementResponse';
import { getPricingSchemeResponseSchema, } from './getPricingSchemeResponse';
import { getSubscriptionResponseSchema, } from './getSubscriptionResponse';
export const getSubscriptionItemResponseSchema = object({
    id: ['id', optional(nullable(string()))],
    description: ['description', optional(nullable(string()))],
    status: ['status', optional(nullable(string()))],
    createdAt: ['created_at', optional(nullable(string()))],
    updatedAt: ['updated_at', optional(nullable(string()))],
    pricingScheme: [
        'pricing_scheme',
        optional(nullable(lazy(() => getPricingSchemeResponseSchema))),
    ],
    discounts: [
        'discounts',
        optional(nullable(array(lazy(() => getDiscountResponseSchema)))),
    ],
    increments: [
        'increments',
        optional(nullable(array(lazy(() => getIncrementResponseSchema)))),
    ],
    subscription: [
        'subscription',
        optional(nullable(lazy(() => getSubscriptionResponseSchema))),
    ],
    name: ['name', optional(nullable(string()))],
    quantity: ['quantity', optional(nullable(number()))],
    cycles: ['cycles', optional(nullable(number()))],
    deletedAt: ['deleted_at', optional(nullable(string()))],
});
