// @pagarme/pagarme-nodejs-sdk/dist/esm/models/getTransfer.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { dict, lazy, number, object, optional, string, } from '../schema';
import { getTransferSourceResponseSchema, } from './getTransferSourceResponse';
import { getTransferTargetResponseSchema, } from './getTransferTargetResponse';
export const getTransferSchema = object({
    id: ['id', string()],
    gatewayId: ['gateway_id', string()],
    amount: ['amount', number()],
    status: ['status', string()],
    createdAt: ['created_at', string()],
    updatedAt: ['updated_at', string()],
    metadata: ['metadata', optional(dict(string()))],
    fee: ['fee', optional(number())],
    fundingDate: ['funding_date', optional(string())],
    fundingEstimatedDate: ['funding_estimated_date', optional(string())],
    type: ['type', string()],
    source: ['source', lazy(() => getTransferSourceResponseSchema)],
    target: ['target', lazy(() => getTransferTargetResponseSchema)],
});
