// @pagarme/pagarme-nodejs-sdk/dist/esm/models/getUsageResponse.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { lazy, nullable, number, object, optional, string, } from '../schema';
import { getSubscriptionItemResponseSchema, } from './getSubscriptionItemResponse';
export const getUsageResponseSchema = object({
    id: ['id', optional(nullable(string()))],
    quantity: ['quantity', optional(nullable(number()))],
    description: ['description', optional(nullable(string()))],
    usedAt: ['used_at', optional(nullable(string()))],
    createdAt: ['created_at', optional(nullable(string()))],
    status: ['status', optional(nullable(string()))],
    deletedAt: ['deleted_at', optional(nullable(string()))],
    subscriptionItem: [
        'subscription_item',
        optional(nullable(lazy(() => getSubscriptionItemResponseSchema))),
    ],
    code: ['code', optional(nullable(string()))],
    group: ['group', optional(nullable(string()))],
    amount: ['amount', optional(nullable(number()))],
});
