// @pagarme/pagarme-nodejs-sdk/dist/esm/models/getWithdrawResponse.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { array, lazy, nullable, number, object, optional, string, } from '../schema';
import { getWithdrawSourceResponseSchema, } from './getWithdrawSourceResponse';
import { getWithdrawTargetResponseSchema, } from './getWithdrawTargetResponse';
export const getWithdrawResponseSchema = object({
    id: ['id', optional(nullable(string()))],
    gatewayId: ['gateway_id', optional(nullable(string()))],
    amount: ['amount', optional(nullable(number()))],
    status: ['status', optional(nullable(string()))],
    createdAt: ['created_at', optional(nullable(string()))],
    updatedAt: ['updated_at', optional(nullable(string()))],
    metadata: ['metadata', optional(nullable(array(string())))],
    fee: ['fee', optional(nullable(number()))],
    fundingDate: ['funding_date', optional(nullable(string()))],
    fundingEstimatedDate: [
        'funding_estimated_date',
        optional(nullable(string())),
    ],
    type: ['type', optional(nullable(string()))],
    source: [
        'source',
        optional(nullable(lazy(() => getWithdrawSourceResponseSchema))),
    ],
    target: [
        'target',
        optional(nullable(lazy(() => getWithdrawTargetResponseSchema))),
    ],
});
