// @pagarme/pagarme-nodejs-sdk/dist/esm/models/updateChargePaymentMethodRequest.js
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
import { boolean, lazy, object, string } from '../schema';
import { createBankTransferPaymentRequestSchema, } from './createBankTransferPaymentRequest';
import { createBoletoPaymentRequestSchema, } from './createBoletoPaymentRequest';
import { createCashPaymentRequestSchema, } from './createCashPaymentRequest';
import { createCreditCardPaymentRequestSchema, } from './createCreditCardPaymentRequest';
import { createDebitCardPaymentRequestSchema, } from './createDebitCardPaymentRequest';
import { createPrivateLabelPaymentRequestSchema, } from './createPrivateLabelPaymentRequest';
import { createVoucherPaymentRequestSchema, } from './createVoucherPaymentRequest';
export const updateChargePaymentMethodRequestSchema = object({
    updateSubscription: ['update_subscription', boolean()],
    paymentMethod: ['payment_method', string()],
    creditCard: [
        'credit_card',
        lazy(() => createCreditCardPaymentRequestSchema),
    ],
    debitCard: ['debit_card', lazy(() => createDebitCardPaymentRequestSchema)],
    boleto: ['boleto', lazy(() => createBoletoPaymentRequestSchema)],
    voucher: ['voucher', lazy(() => createVoucherPaymentRequestSchema)],
    cash: ['cash', lazy(() => createCashPaymentRequestSchema)],
    bankTransfer: [
        'bank_transfer',
        lazy(() => createBankTransferPaymentRequestSchema),
    ],
    privateLabel: [
        'private_label',
        lazy(() => createPrivateLabelPaymentRequestSchema),
    ],
});
