// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/createPrivateLabelPaymentRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.createPrivateLabelPaymentRequestSchema = void 0;
var schema_1 = require("../schema");
var createCardRequest_1 = require("./createCardRequest");
exports.createPrivateLabelPaymentRequestSchema = (0, schema_1.object)({
    installments: ['installments', (0, schema_1.optional)((0, schema_1.number)())],
    statementDescriptor: ['statement_descriptor', (0, schema_1.optional)((0, schema_1.string)())],
    card: ['card', (0, schema_1.optional)((0, schema_1.lazy)(function () { return createCardRequest_1.createCardRequestSchema; }))],
    cardId: ['card_id', (0, schema_1.optional)((0, schema_1.string)())],
    cardToken: ['card_token', (0, schema_1.optional)((0, schema_1.string)())],
    recurrence: ['recurrence', (0, schema_1.optional)((0, schema_1.boolean)())],
    capture: ['capture', (0, schema_1.optional)((0, schema_1.boolean)())],
    extendedLimitEnabled: ['extended_limit_enabled', (0, schema_1.optional)((0, schema_1.boolean)())],
    extendedLimitCode: ['extended_limit_code', (0, schema_1.optional)((0, schema_1.string)())],
    recurrencyCycle: ['recurrency_cycle', (0, schema_1.optional)((0, schema_1.string)())]
});
