// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/createRecipientRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.createRecipientRequestSchema = void 0;
var schema_1 = require("../schema");
var createBankAccountRequest_1 = require("./createBankAccountRequest");
var createRegisterInformationBaseRequest_1 = require("./createRegisterInformationBaseRequest");
var createTransferSettingsRequest_1 = require("./createTransferSettingsRequest");
exports.createRecipientRequestSchema = (0, schema_1.object)({
    name: ['name', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    email: ['email', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    description: ['description', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    document: ['document', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    type: ['type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    defaultBankAccount: [
        'default_bank_account',
        (0, schema_1.lazy)(function () { return createBankAccountRequest_1.createBankAccountRequestSchema; }),
    ],
    metadata: ['metadata', (0, schema_1.dict)((0, schema_1.string)())],
    transferSettings: [
        'transfer_settings',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return createTransferSettingsRequest_1.createTransferSettingsRequestSchema; })),
    ],
    code: ['code', (0, schema_1.string)()],
    paymentMode: ['payment_mode', (0, schema_1.string)()],
    registerInformation: [
        'register_information',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return createRegisterInformationBaseRequest_1.createRegisterInformationBaseRequestSchema; }))),
    ]
});
