// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/createGooglePayRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.createGooglePayRequestSchema = void 0;
var schema_1 = require("../schema");
var createGooglePayIntermediateSigningKeyRequest_1 = require("./createGooglePayIntermediateSigningKeyRequest");
exports.createGooglePayRequestSchema = (0, schema_1.object)({
    version: ['version', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    data: ['data', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    intermediateSigningKey: [
        'intermediate_signing_key',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return createGooglePayIntermediateSigningKeyRequest_1.createGooglePayIntermediateSigningKeyRequestSchema; }))),
    ],
    signature: ['signature', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    signedMessage: ['signed_message', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    merchantIdentifier: ['merchant_identifier', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))]
});
