// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/createCancelChargeRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.createCancelChargeRequestSchema = void 0;
var schema_1 = require("../schema");
var createBankAccountRefundingDTO_1 = require("./createBankAccountRefundingDTO");
var createCancelChargeSplitRulesRequest_1 = require("./createCancelChargeSplitRulesRequest");
var createSplitRequest_1 = require("./createSplitRequest");
exports.createCancelChargeRequestSchema = (0, schema_1.object)({
    amount: ['amount', (0, schema_1.optional)((0, schema_1.number)())],
    splitRules: [
        'split_rules',
        (0, schema_1.optional)((0, schema_1.array)((0, schema_1.lazy)(function () { return createCancelChargeSplitRulesRequest_1.createCancelChargeSplitRulesRequestSchema; }))),
    ],
    split: ['split', (0, schema_1.optional)((0, schema_1.array)((0, schema_1.lazy)(function () { return createSplitRequest_1.createSplitRequestSchema; })))],
    operationReference: ['operation_reference', (0, schema_1.string)()],
    bankAccount: [
        'bank_account',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return createBankAccountRefundingDTO_1.createBankAccountRefundingDTOSchema; })),
    ]
});
