// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/createBankAccountRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.createBankAccountRequestSchema = void 0;
var schema_1 = require("../schema");
exports.createBankAccountRequestSchema = (0, schema_1.object)({
    holderName: ['holder_name', (0, schema_1.string)()],
    holderType: ['holder_type', (0, schema_1.string)()],
    holderDocument: ['holder_document', (0, schema_1.string)()],
    bank: ['bank', (0, schema_1.string)()],
    branchNumber: ['branch_number', (0, schema_1.string)()],
    branchCheckDigit: ['branch_check_digit', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    accountNumber: ['account_number', (0, schema_1.string)()],
    accountCheckDigit: ['account_check_digit', (0, schema_1.string)()],
    type: ['type', (0, schema_1.string)()],
    metadata: ['metadata', (0, schema_1.dict)((0, schema_1.string)())],
    pixKey: ['pix_key', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))]
});
