// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/createSubMerchantRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.createSubMerchantRequestSchema = void 0;
var schema_1 = require("../schema");
var createAddressRequest_1 = require("./createAddressRequest");
var createPhoneRequest_1 = require("./createPhoneRequest");
exports.createSubMerchantRequestSchema = (0, schema_1.object)({
    paymentFacilitatorCode: ['payment_facilitator_code', (0, schema_1.string)()],
    code: ['code', (0, schema_1.string)()],
    name: ['name', (0, schema_1.string)()],
    merchantCategoryCode: ['merchant_category_code', (0, schema_1.string)()],
    document: ['document', (0, schema_1.string)()],
    type: ['type', (0, schema_1.string)()],
    phone: ['phone', (0, schema_1.lazy)(function () { return createPhoneRequest_1.createPhoneRequestSchema; })],
    address: ['address', (0, schema_1.lazy)(function () { return createAddressRequest_1.createAddressRequestSchema; })]
});
