// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/createSubscriptionItemRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.createSubscriptionItemRequestSchema = void 0;
var schema_1 = require("../schema");
var createDiscountRequest_1 = require("./createDiscountRequest");
var createPricingSchemeRequest_1 = require("./createPricingSchemeRequest");
exports.createSubscriptionItemRequestSchema = (0, schema_1.object)({
    description: ['description', (0, schema_1.string)()],
    pricingScheme: [
        'pricing_scheme',
        (0, schema_1.lazy)(function () { return createPricingSchemeRequest_1.createPricingSchemeRequestSchema; }),
    ],
    id: ['id', (0, schema_1.string)()],
    planItemId: ['plan_item_id', (0, schema_1.string)()],
    discounts: ['discounts', (0, schema_1.array)((0, schema_1.lazy)(function () { return createDiscountRequest_1.createDiscountRequestSchema; }))],
    name: ['name', (0, schema_1.string)()],
    cycles: ['cycles', (0, schema_1.optional)((0, schema_1.number)())],
    quantity: ['quantity', (0, schema_1.optional)((0, schema_1.number)())],
    minimumPrice: ['minimum_price', (0, schema_1.optional)((0, schema_1.number)())]
});
