// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/createChargeRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.createChargeRequestSchema = void 0;
var schema_1 = require("../schema");
var createAntifraudRequest_1 = require("./createAntifraudRequest");
var createCustomerRequest_1 = require("./createCustomerRequest");
var createPaymentRequest_1 = require("./createPaymentRequest");
exports.createChargeRequestSchema = (0, schema_1.object)({
    code: ['code', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    amount: ['amount', (0, schema_1.number)()],
    customerId: ['customer_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    customer: [
        'customer',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return createCustomerRequest_1.createCustomerRequestSchema; }))),
    ],
    payment: ['payment', (0, schema_1.lazy)(function () { return createPaymentRequest_1.createPaymentRequestSchema; })],
    metadata: ['metadata', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.dict)((0, schema_1.string)())))],
    dueAt: ['due_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    antifraud: [
        'antifraud',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return createAntifraudRequest_1.createAntifraudRequestSchema; }))),
    ],
    orderId: ['order_id', (0, schema_1.string)()]
});
