// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/createRegisterInformationCorporationRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.createRegisterInformationCorporationRequestSchema = void 0;
var schema_1 = require("../schema");
var createManagingPartnerRequest_1 = require("./createManagingPartnerRequest");
var createRegisterInformationAddressRequest_1 = require("./createRegisterInformationAddressRequest");
var createRegisterInformationPhoneRequest_1 = require("./createRegisterInformationPhoneRequest");
exports.createRegisterInformationCorporationRequestSchema = (0, schema_1.object)({
    companyName: ['company_name', (0, schema_1.string)()],
    tradingName: ['trading_name', (0, schema_1.string)()],
    annualRevenue: ['annual_revenue', (0, schema_1.bigint)()],
    corporationType: ['corporation_type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    foundingDate: ['founding_date', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    cnae: ['cnae', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    managingPartners: [
        'managing_partners',
        (0, schema_1.array)((0, schema_1.lazy)(function () { return createManagingPartnerRequest_1.createManagingPartnerRequestSchema; })),
    ],
    mainAddress: [
        'main_address',
        (0, schema_1.lazy)(function () { return createRegisterInformationAddressRequest_1.createRegisterInformationAddressRequestSchema; }),
    ],
    email: ['email', (0, schema_1.string)()],
    document: ['document', (0, schema_1.string)()],
    type: ['type', (0, schema_1.string)()],
    siteUrl: ['site_url', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    phoneNumbers: [
        'phone_numbers',
        (0, schema_1.array)((0, schema_1.lazy)(function () { return createRegisterInformationPhoneRequest_1.createRegisterInformationPhoneRequestSchema; })),
    ]
});
