// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/createDebitCardPaymentRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.createDebitCardPaymentRequestSchema = void 0;
var schema_1 = require("../schema");
var createCardPaymentContactlessRequest_1 = require("./createCardPaymentContactlessRequest");
var createCardRequest_1 = require("./createCardRequest");
var createPaymentAuthenticationRequest_1 = require("./createPaymentAuthenticationRequest");
var createPaymentOriginRequest_1 = require("./createPaymentOriginRequest");
exports.createDebitCardPaymentRequestSchema = (0, schema_1.object)({
    statementDescriptor: ['statement_descriptor', (0, schema_1.optional)((0, schema_1.string)())],
    card: ['card', (0, schema_1.optional)((0, schema_1.lazy)(function () { return createCardRequest_1.createCardRequestSchema; }))],
    cardId: ['card_id', (0, schema_1.optional)((0, schema_1.string)())],
    cardToken: ['card_token', (0, schema_1.optional)((0, schema_1.string)())],
    recurrence: ['recurrence', (0, schema_1.optional)((0, schema_1.boolean)())],
    authentication: [
        'authentication',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return createPaymentAuthenticationRequest_1.createPaymentAuthenticationRequestSchema; })),
    ],
    token: [
        'token',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return createCardPaymentContactlessRequest_1.createCardPaymentContactlessRequestSchema; })),
    ],
    initiatedType: ['initiated_type', (0, schema_1.optional)((0, schema_1.string)())],
    recurrenceModel: ['recurrence_model', (0, schema_1.optional)((0, schema_1.string)())],
    paymentOrigin: [
        'payment_origin',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return createPaymentOriginRequest_1.createPaymentOriginRequestSchema; })),
    ]
});
