// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/createRegisterInformationIndividualRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.createRegisterInformationIndividualRequestSchema = void 0;
var schema_1 = require("../schema");
var createRegisterInformationAddressRequest_1 = require("./createRegisterInformationAddressRequest");
var createRegisterInformationPhoneRequest_1 = require("./createRegisterInformationPhoneRequest");
exports.createRegisterInformationIndividualRequestSchema = (0, schema_1.object)({
    name: ['name', (0, schema_1.string)()],
    motherName: ['mother_name', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    birthdate: ['birthdate', (0, schema_1.string)()],
    monthlyIncome: ['monthly_income', (0, schema_1.bigint)()],
    professionalOccupation: ['professional_occupation', (0, schema_1.string)()],
    address: [
        'address',
        (0, schema_1.lazy)(function () { return createRegisterInformationAddressRequest_1.createRegisterInformationAddressRequestSchema; }),
    ],
    email: ['email', (0, schema_1.string)()],
    document: ['document', (0, schema_1.string)()],
    type: ['type', (0, schema_1.string)()],
    siteUrl: ['site_url', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    phoneNumbers: [
        'phone_numbers',
        (0, schema_1.array)((0, schema_1.lazy)(function () { return createRegisterInformationPhoneRequest_1.createRegisterInformationPhoneRequestSchema; })),
    ]
});
