// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/createPaymentRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.createPaymentRequestSchema = void 0;
var schema_1 = require("../schema");
var createBankTransferPaymentRequest_1 = require("./createBankTransferPaymentRequest");
var createBoletoPaymentRequest_1 = require("./createBoletoPaymentRequest");
var createCashPaymentRequest_1 = require("./createCashPaymentRequest");
var createCheckoutPaymentRequest_1 = require("./createCheckoutPaymentRequest");
var createCreditCardPaymentRequest_1 = require("./createCreditCardPaymentRequest");
var createCustomerRequest_1 = require("./createCustomerRequest");
var createDebitCardPaymentRequest_1 = require("./createDebitCardPaymentRequest");
var createPixPaymentRequest_1 = require("./createPixPaymentRequest");
var createPrivateLabelPaymentRequest_1 = require("./createPrivateLabelPaymentRequest");
var createSplitRequest_1 = require("./createSplitRequest");
var createVoucherPaymentRequest_1 = require("./createVoucherPaymentRequest");
exports.createPaymentRequestSchema = (0, schema_1.object)({
    paymentMethod: ['payment_method', (0, schema_1.string)()],
    creditCard: [
        'credit_card',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return createCreditCardPaymentRequest_1.createCreditCardPaymentRequestSchema; })),
    ],
    debitCard: [
        'debit_card',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return createDebitCardPaymentRequest_1.createDebitCardPaymentRequestSchema; })),
    ],
    boleto: ['boleto', (0, schema_1.optional)((0, schema_1.lazy)(function () { return createBoletoPaymentRequest_1.createBoletoPaymentRequestSchema; }))],
    currency: ['currency', (0, schema_1.optional)((0, schema_1.string)())],
    voucher: ['voucher', (0, schema_1.optional)((0, schema_1.lazy)(function () { return createVoucherPaymentRequest_1.createVoucherPaymentRequestSchema; }))],
    split: ['split', (0, schema_1.optional)((0, schema_1.array)((0, schema_1.lazy)(function () { return createSplitRequest_1.createSplitRequestSchema; })))],
    bankTransfer: [
        'bank_transfer',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return createBankTransferPaymentRequest_1.createBankTransferPaymentRequestSchema; })),
    ],
    gatewayAffiliationId: ['gateway_affiliation_id', (0, schema_1.optional)((0, schema_1.string)())],
    amount: ['amount', (0, schema_1.optional)((0, schema_1.number)())],
    checkout: [
        'checkout',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return createCheckoutPaymentRequest_1.createCheckoutPaymentRequestSchema; })),
    ],
    customerId: ['customer_id', (0, schema_1.optional)((0, schema_1.string)())],
    customer: ['customer', (0, schema_1.optional)((0, schema_1.lazy)(function () { return createCustomerRequest_1.createCustomerRequestSchema; }))],
    metadata: ['metadata', (0, schema_1.optional)((0, schema_1.dict)((0, schema_1.string)()))],
    cash: ['cash', (0, schema_1.optional)((0, schema_1.lazy)(function () { return createCashPaymentRequest_1.createCashPaymentRequestSchema; }))],
    privateLabel: [
        'private_label',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return createPrivateLabelPaymentRequest_1.createPrivateLabelPaymentRequestSchema; })),
    ],
    pix: ['pix', (0, schema_1.optional)((0, schema_1.lazy)(function () { return createPixPaymentRequest_1.createPixPaymentRequestSchema; }))]
});
