// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/createBoletoPaymentRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.createBoletoPaymentRequestSchema = void 0;
var schema_1 = require("../schema");
var createAddressRequest_1 = require("./createAddressRequest");
var createFineRequest_1 = require("./createFineRequest");
var createInterestRequest_1 = require("./createInterestRequest");
exports.createBoletoPaymentRequestSchema = (0, schema_1.object)({
    retries: ['retries', (0, schema_1.number)()],
    bank: ['bank', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    instructions: ['instructions', (0, schema_1.string)()],
    dueAt: ['due_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    billingAddress: ['billing_address', (0, schema_1.lazy)(function () { return createAddressRequest_1.createAddressRequestSchema; })],
    billingAddressId: ['billing_address_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    nossoNumero: ['nosso_numero', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    documentNumber: ['document_number', (0, schema_1.string)()],
    statementDescriptor: ['statement_descriptor', (0, schema_1.string)()],
    interest: [
        'interest',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return createInterestRequest_1.createInterestRequestSchema; }))),
    ],
    fine: ['fine', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return createFineRequest_1.createFineRequestSchema; })))],
    maxDaysToPayPastDue: [
        'max_days_to_pay_past_due',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)())),
    ]
});
