// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/createPlanRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.createPlanRequestSchema = void 0;
var schema_1 = require("../schema");
var createPlanItemRequest_1 = require("./createPlanItemRequest");
var createPricingSchemeRequest_1 = require("./createPricingSchemeRequest");
exports.createPlanRequestSchema = (0, schema_1.object)({
    name: ['name', (0, schema_1.string)()],
    description: ['description', (0, schema_1.string)()],
    statementDescriptor: ['statement_descriptor', (0, schema_1.string)()],
    items: ['items', (0, schema_1.array)((0, schema_1.lazy)(function () { return createPlanItemRequest_1.createPlanItemRequestSchema; }))],
    shippable: ['shippable', (0, schema_1.boolean)()],
    paymentMethods: ['payment_methods', (0, schema_1.array)((0, schema_1.string)())],
    installments: ['installments', (0, schema_1.array)((0, schema_1.number)())],
    currency: ['currency', (0, schema_1.string)()],
    interval: ['interval', (0, schema_1.string)()],
    intervalCount: ['interval_count', (0, schema_1.number)()],
    billingDays: ['billing_days', (0, schema_1.array)((0, schema_1.number)())],
    billingType: ['billing_type', (0, schema_1.string)()],
    pricingScheme: [
        'pricing_scheme',
        (0, schema_1.lazy)(function () { return createPricingSchemeRequest_1.createPricingSchemeRequestSchema; }),
    ],
    metadata: ['metadata', (0, schema_1.dict)((0, schema_1.string)())],
    minimumPrice: ['minimum_price', (0, schema_1.optional)((0, schema_1.number)())],
    cycles: ['cycles', (0, schema_1.optional)((0, schema_1.number)())],
    quantity: ['quantity', (0, schema_1.optional)((0, schema_1.number)())],
    trialPeriodDays: ['trial_period_days', (0, schema_1.optional)((0, schema_1.number)())]
});
