// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/createCardRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.createCardRequestSchema = void 0;
var schema_1 = require("../schema");
var createAddressRequest_1 = require("./createAddressRequest");
var createCardOptionsRequest_1 = require("./createCardOptionsRequest");
exports.createCardRequestSchema = (0, schema_1.object)({
    number: ['number', (0, schema_1.optional)((0, schema_1.string)())],
    holderName: ['holder_name', (0, schema_1.optional)((0, schema_1.string)())],
    expMonth: ['exp_month', (0, schema_1.optional)((0, schema_1.number)())],
    expYear: ['exp_year', (0, schema_1.optional)((0, schema_1.number)())],
    cvv: ['cvv', (0, schema_1.optional)((0, schema_1.string)())],
    billingAddress: [
        'billing_address',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return createAddressRequest_1.createAddressRequestSchema; })),
    ],
    brand: ['brand', (0, schema_1.optional)((0, schema_1.string)())],
    billingAddressId: ['billing_address_id', (0, schema_1.optional)((0, schema_1.string)())],
    metadata: ['metadata', (0, schema_1.optional)((0, schema_1.dict)((0, schema_1.string)()))],
    type: ['type', (0, schema_1.optional)((0, schema_1.string)())],
    options: ['options', (0, schema_1.optional)((0, schema_1.lazy)(function () { return createCardOptionsRequest_1.createCardOptionsRequestSchema; }))],
    holderDocument: ['holder_document', (0, schema_1.optional)((0, schema_1.string)())],
    privateLabel: ['private_label', (0, schema_1.optional)((0, schema_1.boolean)())],
    label: ['label', (0, schema_1.optional)((0, schema_1.string)())],
    id: ['id', (0, schema_1.optional)((0, schema_1.string)())],
    token: ['token', (0, schema_1.optional)((0, schema_1.string)())]
});
