// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/createShippingRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.createShippingRequestSchema = void 0;
var schema_1 = require("../schema");
var createAddressRequest_1 = require("./createAddressRequest");
exports.createShippingRequestSchema = (0, schema_1.object)({
    amount: ['amount', (0, schema_1.number)()],
    description: ['description', (0, schema_1.string)()],
    recipientName: ['recipient_name', (0, schema_1.string)()],
    recipientPhone: ['recipient_phone', (0, schema_1.string)()],
    addressId: ['address_id', (0, schema_1.string)()],
    address: ['address', (0, schema_1.lazy)(function () { return createAddressRequest_1.createAddressRequestSchema; })],
    maxDeliveryDate: ['max_delivery_date', (0, schema_1.optional)((0, schema_1.string)())],
    estimatedDeliveryDate: ['estimated_delivery_date', (0, schema_1.optional)((0, schema_1.string)())],
    type: ['type', (0, schema_1.string)()]
});
