// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getAnticipationResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getAnticipationResponseSchema = void 0;
var schema_1 = require("../schema");
var getRecipientResponse_1 = require("./getRecipientResponse");
exports.getAnticipationResponseSchema = (0, schema_1.object)({
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    requestedAmount: ['requested_amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    approvedAmount: ['approved_amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    recipient: [
        'recipient',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getRecipientResponse_1.getRecipientResponseSchema; }))),
    ],
    pgid: ['pgid', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    updatedAt: ['updated_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    paymentDate: ['payment_date', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    timeframe: ['timeframe', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))]
});
