// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getAddressResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getAddressResponseSchema = void 0;
var schema_1 = require("../schema");
var getCustomerResponse_1 = require("./getCustomerResponse");
exports.getAddressResponseSchema = (0, schema_1.object)({
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    street: ['street', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    number: ['number', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    complement: ['complement', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    zipCode: ['zip_code', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    neighborhood: ['neighborhood', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    city: ['city', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    state: ['state', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    country: ['country', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    updatedAt: ['updated_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    customer: [
        'customer',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getCustomerResponse_1.getCustomerResponseSchema; }))),
    ],
    metadata: ['metadata', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.dict)((0, schema_1.string)())))],
    line1: ['line_1', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    line2: ['line_2', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    deletedAt: ['deleted_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))]
});
