// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getBalanceOperationResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getBalanceOperationResponseSchema = void 0;
var discriminatedSchemas_1 = require("../models/discriminatedSchemas");
var schema_1 = require("../schema");
exports.getBalanceOperationResponseSchema = (0, schema_1.object)({
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    balanceAmount: ['balance_amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    balanceOldAmount: ['balance_old_amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    type: ['type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    amount: ['amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    fee: ['fee', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    movementObject: [
        'movement_object',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return discriminatedSchemas_1.discriminatedGetMovementObjectBaseResponseSchema; })),
    ]
});
