// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getBankAccountResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getBankAccountResponseSchema = void 0;
var schema_1 = require("../schema");
var getRecipientResponse_1 = require("./getRecipientResponse");
exports.getBankAccountResponseSchema = (0, schema_1.object)({
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    holderName: ['holder_name', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    holderType: ['holder_type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    bank: ['bank', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    branchNumber: ['branch_number', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    branchCheckDigit: ['branch_check_digit', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    accountNumber: ['account_number', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    accountCheckDigit: ['account_check_digit', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    type: ['type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    updatedAt: ['updated_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    deletedAt: ['deleted_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    recipient: [
        'recipient',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getRecipientResponse_1.getRecipientResponseSchema; }))),
    ],
    metadata: ['metadata', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.dict)((0, schema_1.string)())))],
    pixKey: ['pix_key', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))]
});
