// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getBillingAddressResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getBillingAddressResponseSchema = void 0;
var schema_1 = require("../schema");
exports.getBillingAddressResponseSchema = (0, schema_1.object)({
    street: ['street', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    number: ['number', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    zipCode: ['zip_code', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    neighborhood: ['neighborhood', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    city: ['city', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    state: ['state', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    country: ['country', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    complement: ['complement', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    line1: ['line_1', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    line2: ['line_2', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))]
});
