// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getBalanceResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getBalanceResponseSchema = void 0;
var schema_1 = require("../schema");
var getRecipientResponse_1 = require("./getRecipientResponse");
exports.getBalanceResponseSchema = (0, schema_1.object)({
    currency: ['currency', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    availableAmount: ['available_amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.bigint)()))],
    recipient: [
        'recipient',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getRecipientResponse_1.getRecipientResponseSchema; }))),
    ],
    transferredAmount: ['transferred_amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.bigint)()))],
    waitingFundsAmount: ['waiting_funds_amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.bigint)()))]
});
