// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getCardTokenResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getCardTokenResponseSchema = void 0;
var schema_1 = require("../schema");
exports.getCardTokenResponseSchema = (0, schema_1.object)({
    lastFourDigits: ['last_four_digits', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    holderName: ['holder_name', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    holderDocument: ['holder_document', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    expMonth: ['exp_month', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    expYear: ['exp_year', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    brand: ['brand', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    type: ['type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    label: ['label', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))]
});
