// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getCheckoutPaymentSettingsResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getCheckoutPaymentSettingsResponseSchema = void 0;
var schema_1 = require("../schema");
var getCustomerResponse_1 = require("./getCustomerResponse");
exports.getCheckoutPaymentSettingsResponseSchema = (0, schema_1.object)({
    successUrl: ['success_url', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    paymentUrl: ['payment_url', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    acceptedPaymentMethods: [
        'accepted_payment_methods',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.string)()))),
    ],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    customer: [
        'customer',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getCustomerResponse_1.getCustomerResponseSchema; }))),
    ],
    amount: ['amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    defaultPaymentMethod: [
        'default_payment_method',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)())),
    ],
    gatewayAffiliationId: [
        'gateway_affiliation_id',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)())),
    ]
});
