// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getCardResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getCardResponseSchema = void 0;
var schema_1 = require("../schema");
var getBillingAddressResponse_1 = require("./getBillingAddressResponse");
var getCustomerResponse_1 = require("./getCustomerResponse");
exports.getCardResponseSchema = (0, schema_1.object)({
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    lastFourDigits: ['last_four_digits', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    brand: ['brand', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    holderName: ['holder_name', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    expMonth: ['exp_month', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    expYear: ['exp_year', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    updatedAt: ['updated_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    billingAddress: [
        'billing_address',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getBillingAddressResponse_1.getBillingAddressResponseSchema; }))),
    ],
    customer: [
        'customer',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getCustomerResponse_1.getCustomerResponseSchema; }))),
    ],
    metadata: ['metadata', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.dict)((0, schema_1.string)())))],
    type: ['type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    holderDocument: ['holder_document', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    deletedAt: ['deleted_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    firstSixDigits: ['first_six_digits', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    label: ['label', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))]
});
