// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getCustomerResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getCustomerResponseSchema = void 0;
var schema_1 = require("../schema");
var getAddressResponse_1 = require("./getAddressResponse");
var getPhonesResponse_1 = require("./getPhonesResponse");
exports.getCustomerResponseSchema = (0, schema_1.object)({
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    name: ['name', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    email: ['email', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    delinquent: ['delinquent', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.boolean)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    updatedAt: ['updated_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    document: ['document', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    type: ['type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    fbAccessToken: ['fb_access_token', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    address: [
        'address',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getAddressResponse_1.getAddressResponseSchema; }))),
    ],
    metadata: ['metadata', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.dict)((0, schema_1.string)())))],
    phones: ['phones', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getPhonesResponse_1.getPhonesResponseSchema; })))],
    fbId: ['fb_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.bigint)()))],
    code: ['code', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    documentType: ['document_type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))]
});
