// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getCreditCardTransactionResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getCreditCardTransactionResponseSchema = void 0;
var schema_1 = require("../schema");
var getAntifraudResponse_1 = require("./getAntifraudResponse");
var getCardResponse_1 = require("./getCardResponse");
var getFineResponse_1 = require("./getFineResponse");
var getGatewayResponseResponse_1 = require("./getGatewayResponseResponse");
var getInterestResponse_1 = require("./getInterestResponse");
var getRetryTransactionInformationResponse_1 = require("./getRetryTransactionInformationResponse");
var getSplitResponse_1 = require("./getSplitResponse");
exports.getCreditCardTransactionResponseSchema = (0, schema_1.object)({
    statementDescriptor: ['statement_descriptor', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    acquirerName: ['acquirer_name', (0, schema_1.optional)((0, schema_1.string)())],
    acquirerAffiliationCode: [
        'acquirer_affiliation_code',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)())),
    ],
    acquirerTid: ['acquirer_tid', (0, schema_1.optional)((0, schema_1.string)())],
    acquirerNsu: ['acquirer_nsu', (0, schema_1.optional)((0, schema_1.string)())],
    acquirerAuthCode: ['acquirer_auth_code', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    operationType: ['operation_type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    card: ['card', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getCardResponse_1.getCardResponseSchema; })))],
    acquirerMessage: ['acquirer_message', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    acquirerReturnCode: ['acquirer_return_code', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    installments: ['installments', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    threedAuthenticationUrl: [
        'threed_authentication_url',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)())),
    ],
    fundingSource: ['funding_source', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    retryInfo: [
        'retry_info',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getRetryTransactionInformationResponse_1.getRetryTransactionInformationResponseSchema; }))),
    ],
    brandId: ['brand_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    gatewayId: ['gateway_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    amount: ['amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    success: ['success', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.boolean)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    updatedAt: ['updated_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    attemptCount: ['attempt_count', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    maxAttempts: ['max_attempts', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    splits: [
        'splits',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.lazy)(function () { return getSplitResponse_1.getSplitResponseSchema; })))),
    ],
    nextAttempt: ['next_attempt', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    gatewayResponse: [
        'gateway_response',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getGatewayResponseResponse_1.getGatewayResponseResponseSchema; }))),
    ],
    antifraudResponse: [
        'antifraud_response',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getAntifraudResponse_1.getAntifraudResponseSchema; }))),
    ],
    metadata: ['metadata', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.dict)((0, schema_1.string)())))],
    split: [
        'split',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.lazy)(function () { return getSplitResponse_1.getSplitResponseSchema; })))),
    ],
    interest: [
        'interest',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getInterestResponse_1.getInterestResponseSchema; }))),
    ],
    fine: ['fine', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getFineResponse_1.getFineResponseSchema; })))],
    maxDaysToPayPastDue: [
        'max_days_to_pay_past_due',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)())),
    ],
    transactionType: ['transaction_type', (0, schema_1.optional)((0, schema_1.string)())]
});
