// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getInvoiceItemResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getInvoiceItemResponseSchema = void 0;
var schema_1 = require("../schema");
var getPriceBracketResponse_1 = require("./getPriceBracketResponse");
var getPricingSchemeResponse_1 = require("./getPricingSchemeResponse");
exports.getInvoiceItemResponseSchema = (0, schema_1.object)({
    amount: ['amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    description: ['description', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    pricingScheme: [
        'pricing_scheme',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getPricingSchemeResponse_1.getPricingSchemeResponseSchema; }))),
    ],
    priceBracket: [
        'price_bracket',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getPriceBracketResponse_1.getPriceBracketResponseSchema; }))),
    ],
    quantity: ['quantity', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    name: ['name', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    subscriptionItemId: ['subscription_item_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))]
});
