// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getInvoiceResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getInvoiceResponseSchema = void 0;
var schema_1 = require("../schema");
var getBillingAddressResponse_1 = require("./getBillingAddressResponse");
var getChargeResponse_1 = require("./getChargeResponse");
var getCustomerResponse_1 = require("./getCustomerResponse");
var getInvoiceItemResponse_1 = require("./getInvoiceItemResponse");
var getPeriodResponse_1 = require("./getPeriodResponse");
var getShippingResponse_1 = require("./getShippingResponse");
var getSubscriptionResponse_1 = require("./getSubscriptionResponse");
exports.getInvoiceResponseSchema = (0, schema_1.object)({
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    code: ['code', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    url: ['url', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    amount: ['amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    paymentMethod: ['payment_method', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    items: [
        'items',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.lazy)(function () { return getInvoiceItemResponse_1.getInvoiceItemResponseSchema; })))),
    ],
    customer: [
        'customer',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getCustomerResponse_1.getCustomerResponseSchema; }))),
    ],
    charge: ['charge', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getChargeResponse_1.getChargeResponseSchema; })))],
    installments: ['installments', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    billingAddress: [
        'billing_address',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getBillingAddressResponse_1.getBillingAddressResponseSchema; }))),
    ],
    subscription: [
        'subscription',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getSubscriptionResponse_1.getSubscriptionResponseSchema; }))),
    ],
    cycle: ['cycle', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getPeriodResponse_1.getPeriodResponseSchema; })))],
    shipping: [
        'shipping',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getShippingResponse_1.getShippingResponseSchema; }))),
    ],
    metadata: ['metadata', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.dict)((0, schema_1.string)())))],
    dueAt: ['due_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    canceledAt: ['canceled_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    billingAt: ['billing_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    seenAt: ['seen_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    totalDiscount: ['total_discount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    totalIncrement: ['total_increment', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    subscriptionId: ['subscription_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))]
});
