// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getIncrementResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getIncrementResponseSchema = void 0;
var schema_1 = require("../schema");
var getSubscriptionItemResponse_1 = require("./getSubscriptionItemResponse");
var getSubscriptionResponse_1 = require("./getSubscriptionResponse");
exports.getIncrementResponseSchema = (0, schema_1.object)({
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    value: ['value', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    incrementType: ['increment_type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    cycles: ['cycles', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    deletedAt: ['deleted_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    description: ['description', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    subscription: [
        'subscription',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getSubscriptionResponse_1.getSubscriptionResponseSchema; }))),
    ],
    subscriptionItem: [
        'subscription_item',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getSubscriptionItemResponse_1.getSubscriptionItemResponseSchema; }))),
    ]
});
