// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getMovementObjectFeeCollectionResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getMovementObjectFeeCollectionResponseSchema = void 0;
var schema_1 = require("../schema");
exports.getMovementObjectFeeCollectionResponseSchema = (0, schema_1.object)({
    description: ['description', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    paymentDate: ['payment_date', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    recipientId: ['recipient_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    amount: ['amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    type: ['type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    chargeId: ['charge_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    gatewayId: ['gateway_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    object: ['object', (0, schema_1.optional)((0, schema_1.string)())]
});
