// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getManagingPartnerResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getManagingPartnerResponseSchema = void 0;
var schema_1 = require("../schema");
var getPhoneNumberResponse_1 = require("./getPhoneNumberResponse");
var getRegisterInformationAddressResponse_1 = require("./getRegisterInformationAddressResponse");
exports.getManagingPartnerResponseSchema = (0, schema_1.object)({
    name: ['name', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    email: ['email', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    document: ['document', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    type: ['type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    motherName: ['mother_name', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    birthdate: ['birthdate', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    monthlyIncome: ['monthly_income', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    professionalOccupation: [
        'professional_occupation',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)())),
    ],
    selfDeclaredRepresentative: [
        'self_declared_representative',
        (0, schema_1.optional)((0, schema_1.boolean)()),
    ],
    address: [
        'address',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getRegisterInformationAddressResponse_1.getRegisterInformationAddressResponseSchema; }))),
    ],
    phoneNumbers: [
        'phone_numbers',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.lazy)(function () { return getPhoneNumberResponse_1.getPhoneNumberResponseSchema; })))),
    ]
});
