// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getMovementObjectRefundResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getMovementObjectRefundResponseSchema = void 0;
var schema_1 = require("../schema");
exports.getMovementObjectRefundResponseSchema = (0, schema_1.object)({
    fraudCoverageFee: ['fraud_coverage_fee', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    chargeFeeRecipientId: [
        'charge_fee_recipient_id',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)())),
    ],
    bankAccountId: ['bank_account_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    localTransactionId: ['local_transaction_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    updatedAt: ['updated_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    amount: ['amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    type: ['type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    chargeId: ['charge_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    gatewayId: ['gateway_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    object: ['object', (0, schema_1.optional)((0, schema_1.string)())]
});
