// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getMovementObjectSettlementResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getMovementObjectSettlementResponseSchema = void 0;
var schema_1 = require("../schema");
exports.getMovementObjectSettlementResponseSchema = (0, schema_1.object)({
    product: ['product', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    brand: ['brand', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    paymentDate: ['payment_date', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    recipientId: ['recipient_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    documentType: ['document_type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    document: ['document', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    contractObligationId: [
        'contract_obligation_id',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)())),
    ],
    liquidationArrangementId: [
        'liquidation_arrangement_id',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)())),
    ],
    externalEnginePaymentId: [
        'external_engine_payment_id',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)())),
    ],
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    amount: ['amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    type: ['type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    chargeId: ['charge_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    gatewayId: ['gateway_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    object: ['object', (0, schema_1.optional)((0, schema_1.string)())]
});
