// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getMovementObjectTransferResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getMovementObjectTransferResponseSchema = void 0;
var schema_1 = require("../schema");
exports.getMovementObjectTransferResponseSchema = (0, schema_1.object)({
    sourceType: ['source_type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    sourceId: ['source_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    targetType: ['target_type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    targetId: ['target_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    fee: ['fee', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    fundingDate: ['funding_date', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    fundingEstimatedDate: [
        'funding_estimated_date',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)())),
    ],
    bankAccount: ['bank_account', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    amount: ['amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    type: ['type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    chargeId: ['charge_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    gatewayId: ['gateway_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    object: ['object', (0, schema_1.optional)((0, schema_1.string)())]
});
