// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getPeriodResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getPeriodResponseSchema = void 0;
var schema_1 = require("../schema");
var getSubscriptionResponse_1 = require("./getSubscriptionResponse");
exports.getPeriodResponseSchema = (0, schema_1.object)({
    startAt: ['start_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    endAt: ['end_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    billingAt: ['billing_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    subscription: [
        'subscription',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getSubscriptionResponse_1.getSubscriptionResponseSchema; }))),
    ],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    duration: ['duration', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    updatedAt: ['updated_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    cycle: ['cycle', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))]
});
