// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getPlanItemResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getPlanItemResponseSchema = void 0;
var schema_1 = require("../schema");
var getPlanResponse_1 = require("./getPlanResponse");
var getPricingSchemeResponse_1 = require("./getPricingSchemeResponse");
exports.getPlanItemResponseSchema = (0, schema_1.object)({
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    name: ['name', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    updatedAt: ['updated_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    pricingScheme: [
        'pricing_scheme',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getPricingSchemeResponse_1.getPricingSchemeResponseSchema; }))),
    ],
    description: ['description', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    plan: ['plan', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getPlanResponse_1.getPlanResponseSchema; })))],
    quantity: ['quantity', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    cycles: ['cycles', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    deletedAt: ['deleted_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))]
});
