// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getPlanResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getPlanResponseSchema = void 0;
var schema_1 = require("../schema");
var getPlanItemResponse_1 = require("./getPlanItemResponse");
exports.getPlanResponseSchema = (0, schema_1.object)({
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    name: ['name', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    description: ['description', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    url: ['url', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    statementDescriptor: ['statement_descriptor', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    interval: ['interval', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    intervalCount: ['interval_count', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    billingType: ['billing_type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    paymentMethods: ['payment_methods', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.string)())))],
    installments: ['installments', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.number)())))],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    currency: ['currency', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    updatedAt: ['updated_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    items: [
        'items',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.lazy)(function () { return getPlanItemResponse_1.getPlanItemResponseSchema; })))),
    ],
    billingDays: ['billing_days', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.number)())))],
    shippable: ['shippable', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.boolean)()))],
    metadata: ['metadata', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.dict)((0, schema_1.string)())))],
    trialPeriodDays: ['trial_period_days', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    minimumPrice: ['minimum_price', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    deletedAt: ['deleted_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))]
});
