// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getPricingSchemeResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getPricingSchemeResponseSchema = void 0;
var schema_1 = require("../schema");
var getPriceBracketResponse_1 = require("./getPriceBracketResponse");
exports.getPricingSchemeResponseSchema = (0, schema_1.object)({
    price: ['price', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    schemeType: ['scheme_type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    priceBrackets: [
        'price_brackets',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.lazy)(function () { return getPriceBracketResponse_1.getPriceBracketResponseSchema; })))),
    ],
    minimumPrice: ['minimum_price', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    percentage: ['percentage', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))]
});
