// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getPayableResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getPayableResponseSchema = void 0;
var schema_1 = require("../schema");
exports.getPayableResponseSchema = (0, schema_1.object)({
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.bigint)()))],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    amount: ['amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    fee: ['fee', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    anticipationFee: ['anticipation_fee', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    fraudCoverageFee: ['fraud_coverage_fee', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    installment: ['installment', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    gatewayId: ['gateway_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.bigint)()))],
    chargeId: ['charge_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    splitId: ['split_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    bulkAnticipationId: ['bulk_anticipation_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    anticipationId: ['anticipation_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    recipientId: ['recipient_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    originatorModel: ['originator_model', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    originatorModelId: ['originator_model_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    paymentDate: ['payment_date', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    originalPaymentDate: ['original_payment_date', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    type: ['type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    paymentMethod: ['payment_method', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    accrualAt: ['accrual_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    liquidationArrangementId: [
        'liquidation_arrangement_id',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)())),
    ]
});
