// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getRegisterInformationAddressResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getRegisterInformationAddressResponseSchema = void 0;
var schema_1 = require("../schema");
exports.getRegisterInformationAddressResponseSchema = (0, schema_1.object)({
    street: ['street', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    complementary: ['complementary', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    streetNumber: ['street_number', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    neighborhood: ['neighborhood', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    city: ['city', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    state: ['state', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    zipCode: ['zip_code', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    referencePoint: ['reference_point', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))]
});
