// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getRecipientResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getRecipientResponseSchema = void 0;
var schema_1 = require("../schema");
var getAutomaticAnticipationResponse_1 = require("./getAutomaticAnticipationResponse");
var getBankAccountResponse_1 = require("./getBankAccountResponse");
var getGatewayRecipientResponse_1 = require("./getGatewayRecipientResponse");
var getRegisterInformationResponse_1 = require("./getRegisterInformationResponse");
var getTransferSettingsResponse_1 = require("./getTransferSettingsResponse");
exports.getRecipientResponseSchema = (0, schema_1.object)({
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    name: ['name', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    email: ['email', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    document: ['document', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    description: ['description', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    type: ['type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    updatedAt: ['updated_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    deletedAt: ['deleted_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    defaultBankAccount: [
        'default_bank_account',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getBankAccountResponse_1.getBankAccountResponseSchema; }))),
    ],
    gatewayRecipients: [
        'gateway_recipients',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.lazy)(function () { return getGatewayRecipientResponse_1.getGatewayRecipientResponseSchema; })))),
    ],
    metadata: ['metadata', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.dict)((0, schema_1.string)())))],
    automaticAnticipationSettings: [
        'automatic_anticipation_settings',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getAutomaticAnticipationResponse_1.getAutomaticAnticipationResponseSchema; }))),
    ],
    transferSettings: [
        'transfer_settings',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getTransferSettingsResponse_1.getTransferSettingsResponseSchema; }))),
    ],
    code: ['code', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    paymentMode: ['payment_mode', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    registerInformation: [
        'register_information',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getRegisterInformationResponse_1.getRegisterInformationResponseSchema; }))),
    ]
});
