// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getSplitResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getSplitResponseSchema = void 0;
var schema_1 = require("../schema");
var getRecipientResponse_1 = require("./getRecipientResponse");
var getSplitOptionsResponse_1 = require("./getSplitOptionsResponse");
exports.getSplitResponseSchema = (0, schema_1.object)({
    type: ['type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    amount: ['amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    recipient: [
        'recipient',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getRecipientResponse_1.getRecipientResponseSchema; }))),
    ],
    gatewayId: ['gateway_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    options: [
        'options',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getSplitOptionsResponse_1.getSplitOptionsResponseSchema; }))),
    ],
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))]
});
