// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getShippingResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getShippingResponseSchema = void 0;
var schema_1 = require("../schema");
var getAddressResponse_1 = require("./getAddressResponse");
exports.getShippingResponseSchema = (0, schema_1.object)({
    amount: ['amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    description: ['description', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    recipientName: ['recipient_name', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    recipientPhone: ['recipient_phone', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    address: [
        'address',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getAddressResponse_1.getAddressResponseSchema; }))),
    ],
    maxDeliveryDate: ['max_delivery_date', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    estimatedDeliveryDate: [
        'estimated_delivery_date',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)())),
    ],
    type: ['type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))]
});
