// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getSubscriptionItemResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getSubscriptionItemResponseSchema = void 0;
var schema_1 = require("../schema");
var getDiscountResponse_1 = require("./getDiscountResponse");
var getIncrementResponse_1 = require("./getIncrementResponse");
var getPricingSchemeResponse_1 = require("./getPricingSchemeResponse");
var getSubscriptionResponse_1 = require("./getSubscriptionResponse");
exports.getSubscriptionItemResponseSchema = (0, schema_1.object)({
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    description: ['description', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    updatedAt: ['updated_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    pricingScheme: [
        'pricing_scheme',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getPricingSchemeResponse_1.getPricingSchemeResponseSchema; }))),
    ],
    discounts: [
        'discounts',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.lazy)(function () { return getDiscountResponse_1.getDiscountResponseSchema; })))),
    ],
    increments: [
        'increments',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.lazy)(function () { return getIncrementResponse_1.getIncrementResponseSchema; })))),
    ],
    subscription: [
        'subscription',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getSubscriptionResponse_1.getSubscriptionResponseSchema; }))),
    ],
    name: ['name', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    quantity: ['quantity', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    cycles: ['cycles', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    deletedAt: ['deleted_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))]
});
