// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getTransfer.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getTransferSchema = void 0;
var schema_1 = require("../schema");
var getTransferSourceResponse_1 = require("./getTransferSourceResponse");
var getTransferTargetResponse_1 = require("./getTransferTargetResponse");
exports.getTransferSchema = (0, schema_1.object)({
    id: ['id', (0, schema_1.string)()],
    gatewayId: ['gateway_id', (0, schema_1.string)()],
    amount: ['amount', (0, schema_1.number)()],
    status: ['status', (0, schema_1.string)()],
    createdAt: ['created_at', (0, schema_1.string)()],
    updatedAt: ['updated_at', (0, schema_1.string)()],
    metadata: ['metadata', (0, schema_1.optional)((0, schema_1.dict)((0, schema_1.string)()))],
    fee: ['fee', (0, schema_1.optional)((0, schema_1.number)())],
    fundingDate: ['funding_date', (0, schema_1.optional)((0, schema_1.string)())],
    fundingEstimatedDate: ['funding_estimated_date', (0, schema_1.optional)((0, schema_1.string)())],
    type: ['type', (0, schema_1.string)()],
    source: ['source', (0, schema_1.lazy)(function () { return getTransferSourceResponse_1.getTransferSourceResponseSchema; })],
    target: ['target', (0, schema_1.lazy)(function () { return getTransferTargetResponse_1.getTransferTargetResponseSchema; })]
});
