// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getTransferResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getTransferResponseSchema = void 0;
var schema_1 = require("../schema");
var getBankAccountResponse_1 = require("./getBankAccountResponse");
exports.getTransferResponseSchema = (0, schema_1.object)({
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    amount: ['amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    updatedAt: ['updated_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    bankAccount: [
        'bank_account',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getBankAccountResponse_1.getBankAccountResponseSchema; }))),
    ],
    metadata: ['metadata', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.dict)((0, schema_1.string)())))]
});
