// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/getWithdrawResponse.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.getWithdrawResponseSchema = void 0;
var schema_1 = require("../schema");
var getWithdrawSourceResponse_1 = require("./getWithdrawSourceResponse");
var getWithdrawTargetResponse_1 = require("./getWithdrawTargetResponse");
exports.getWithdrawResponseSchema = (0, schema_1.object)({
    id: ['id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    gatewayId: ['gateway_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    amount: ['amount', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    status: ['status', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    updatedAt: ['updated_at', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    metadata: ['metadata', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.string)())))],
    fee: ['fee', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    fundingDate: ['funding_date', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    fundingEstimatedDate: [
        'funding_estimated_date',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)())),
    ],
    type: ['type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    source: [
        'source',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getWithdrawSourceResponse_1.getWithdrawSourceResponseSchema; }))),
    ],
    target: [
        'target',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.lazy)(function () { return getWithdrawTargetResponse_1.getWithdrawTargetResponseSchema; }))),
    ]
});
