// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/updateCustomerRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.updateCustomerRequestSchema = void 0;
var schema_1 = require("../schema");
var createAddressRequest_1 = require("./createAddressRequest");
var createPhonesRequest_1 = require("./createPhonesRequest");
exports.updateCustomerRequestSchema = (0, schema_1.object)({
    name: ['name', (0, schema_1.optional)((0, schema_1.string)())],
    email: ['email', (0, schema_1.optional)((0, schema_1.string)())],
    document: ['document', (0, schema_1.optional)((0, schema_1.string)())],
    type: ['type', (0, schema_1.optional)((0, schema_1.string)())],
    address: ['address', (0, schema_1.optional)((0, schema_1.lazy)(function () { return createAddressRequest_1.createAddressRequestSchema; }))],
    metadata: ['metadata', (0, schema_1.optional)((0, schema_1.dict)((0, schema_1.string)()))],
    phones: ['phones', (0, schema_1.optional)((0, schema_1.lazy)(function () { return createPhonesRequest_1.createPhonesRequestSchema; }))],
    code: ['code', (0, schema_1.optional)((0, schema_1.string)())],
    gender: ['gender', (0, schema_1.optional)((0, schema_1.string)())],
    documentType: ['document_type', (0, schema_1.optional)((0, schema_1.string)())]
});
