// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/updateChargeCardRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.updateChargeCardRequestSchema = void 0;
var schema_1 = require("../schema");
var createCardRequest_1 = require("./createCardRequest");
var createPaymentOriginRequest_1 = require("./createPaymentOriginRequest");
exports.updateChargeCardRequestSchema = (0, schema_1.object)({
    updateSubscription: ['update_subscription', (0, schema_1.boolean)()],
    cardId: ['card_id', (0, schema_1.string)()],
    card: ['card', (0, schema_1.lazy)(function () { return createCardRequest_1.createCardRequestSchema; })],
    recurrence: ['recurrence', (0, schema_1.boolean)()],
    initiatedType: ['initiated_type', (0, schema_1.optional)((0, schema_1.string)())],
    recurrenceModel: ['recurrence_model', (0, schema_1.optional)((0, schema_1.string)())],
    paymentOrigin: [
        'payment_origin',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return createPaymentOriginRequest_1.createPaymentOriginRequestSchema; })),
    ]
});
