// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/updateChargePaymentMethodRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.updateChargePaymentMethodRequestSchema = void 0;
var schema_1 = require("../schema");
var createBankTransferPaymentRequest_1 = require("./createBankTransferPaymentRequest");
var createBoletoPaymentRequest_1 = require("./createBoletoPaymentRequest");
var createCashPaymentRequest_1 = require("./createCashPaymentRequest");
var createCreditCardPaymentRequest_1 = require("./createCreditCardPaymentRequest");
var createDebitCardPaymentRequest_1 = require("./createDebitCardPaymentRequest");
var createPrivateLabelPaymentRequest_1 = require("./createPrivateLabelPaymentRequest");
var createVoucherPaymentRequest_1 = require("./createVoucherPaymentRequest");
exports.updateChargePaymentMethodRequestSchema = (0, schema_1.object)({
    updateSubscription: ['update_subscription', (0, schema_1.boolean)()],
    paymentMethod: ['payment_method', (0, schema_1.string)()],
    creditCard: [
        'credit_card',
        (0, schema_1.lazy)(function () { return createCreditCardPaymentRequest_1.createCreditCardPaymentRequestSchema; }),
    ],
    debitCard: ['debit_card', (0, schema_1.lazy)(function () { return createDebitCardPaymentRequest_1.createDebitCardPaymentRequestSchema; })],
    boleto: ['boleto', (0, schema_1.lazy)(function () { return createBoletoPaymentRequest_1.createBoletoPaymentRequestSchema; })],
    voucher: ['voucher', (0, schema_1.lazy)(function () { return createVoucherPaymentRequest_1.createVoucherPaymentRequestSchema; })],
    cash: ['cash', (0, schema_1.lazy)(function () { return createCashPaymentRequest_1.createCashPaymentRequestSchema; })],
    bankTransfer: [
        'bank_transfer',
        (0, schema_1.lazy)(function () { return createBankTransferPaymentRequest_1.createBankTransferPaymentRequestSchema; }),
    ],
    privateLabel: [
        'private_label',
        (0, schema_1.lazy)(function () { return createPrivateLabelPaymentRequest_1.createPrivateLabelPaymentRequestSchema; }),
    ]
});
