// @pagarme/pagarme-nodejs-sdk/dist/cjs/models/updatePlanRequest.js
"use strict";
/**
 * PagarmeApiSDKLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */
exports.__esModule = true;
exports.updatePlanRequestSchema = void 0;
var schema_1 = require("../schema");
exports.updatePlanRequestSchema = (0, schema_1.object)({
    name: ['name', (0, schema_1.string)()],
    description: ['description', (0, schema_1.string)()],
    installments: ['installments', (0, schema_1.array)((0, schema_1.number)())],
    statementDescriptor: ['statement_descriptor', (0, schema_1.string)()],
    currency: ['currency', (0, schema_1.string)()],
    interval: ['interval', (0, schema_1.string)()],
    intervalCount: ['interval_count', (0, schema_1.number)()],
    paymentMethods: ['payment_methods', (0, schema_1.array)((0, schema_1.string)())],
    billingType: ['billing_type', (0, schema_1.string)()],
    status: ['status', (0, schema_1.string)()],
    shippable: ['shippable', (0, schema_1.boolean)()],
    billingDays: ['billing_days', (0, schema_1.array)((0, schema_1.number)())],
    metadata: ['metadata', (0, schema_1.dict)((0, schema_1.string)())],
    minimumPrice: ['minimum_price', (0, schema_1.optional)((0, schema_1.number)())],
    trialPeriodDays: ['trial_period_days', (0, schema_1.optional)((0, schema_1.number)())]
});
